File: main/src/main/java/lavaplayer/source/youtube/DefaultYoutubePlaylistLoader.java
package lavaplayer.source.youtube;

import lavaplayer.tools.FriendlyException;
import lavaplayer.tools.JsonBrowser;
import lavaplayer.tools.ThumbnailTools;
import lavaplayer.tools.Units;
import lavaplayer.tools.io.HttpClientTools;
import lavaplayer.tools.io.HttpInterface;
import lavaplayer.track.*;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.StringEntity;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.function.Function;
import java.util.stream.Collectors;

import static lavaplayer.tools.FriendlyException.Severity.COMMON;

public class DefaultYoutubePlaylistLoader implements YoutubePlaylistLoader {
    private volatile int playlistPageCount = 6;

    @Override
    public void setPlaylistPageCount(int playlistPageCount) {
        this.playlistPageCount = playlistPageCount;
    }

    @Override
    public BasicAudioTrackCollection load(HttpInterface httpInterface, String playlistId, String selectedVideoId,
                              Function<AudioTrackInfo, AudioTrack> trackFactory) {
        HttpPost post = new HttpPost(YoutubeConstants.BROWSE_URL);
        StringEntity payload = new StringEntity(String.format(YoutubeConstants.BROWSE_PLAYLIST_PAYLOAD, playlistId), "UTF-8");
        post.setEntity(payload);
        try (CloseableHttpResponse response = httpInterface.execute(post)) {
            HttpClientTools.assertSuccessWithContent(response, "playlist response");
            HttpClientTools.assertJsonContentType(response);

            JsonBrowser json = JsonBrowser.parse(response.getEntity().getContent());
            return buildPlaylist(httpInterface, json, selectedVideoId, trackFactory);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private BasicAudioTrackCollection buildPlaylist(HttpInterface httpInterface, JsonBrowser json, String selectedVideoId,
                                        Function<AudioTrackInfo, AudioTrack> trackFactory) throws IOException {

        String errorAlertMessage = findErrorAlert(json);

        if (errorAlertMessage != null) {
            throw new FriendlyException(errorAlertMessage, COMMON, null);
        }

        String playlistName = json
            .get("header")
            .get("playlistHeaderRenderer")
            .get("title")
            .get("runs")
            .index(0)
            .get("text")
            .text();

        JsonBrowser playlistVideoList = json
            .get("contents")
            .get("singleColumnBrowseResultsRenderer")
            .get("tabs")
            .index(0)
            .get("tabRenderer")
            .get("content")
            .get("sectionListRenderer")
            .get("contents")
            .index(0)
            .get("playlistVideoListRenderer");

        List<AudioTrack> tracks = new ArrayList<>();
        String continuationsToken = extractPlaylistTracks(playlistVideoList, tracks, trackFactory);
        int loadCount = 0;
        int pageCount = playlistPageCount;

        // Also load the next pages, each result gives us a JSON with separate values for list html and next page loader html
        while (continuationsToken != null && ++loadCount < pageCount) {
            HttpPost post = new HttpPost(YoutubeConstants.BROWSE_URL);
            StringEntity payload = new StringEntity(String.format(YoutubeConstants.BROWSE_CONTINUATION_PAYLOAD, continuationsToken), "UTF-8");
            post.setEntity(payload);
            try (CloseableHttpResponse response = httpInterface.execute(post)) {
                HttpClientTools.assertSuccessWithContent(response, "playlist response");

                JsonBrowser continuationJson = JsonBrowser.parse(response.getEntity().getContent());

                JsonBrowser playlistVideoListPage = continuationJson
                    .get("continuationContents")
                    .get("playlistVideoListContinuation");

                continuationsToken = extractPlaylistTracks(playlistVideoListPage, tracks, trackFactory);
            }
        }

        return new BasicAudioTrackCollection(playlistName, AudioTrackCollectionType.Playlist.INSTANCE, tracks, findSelectedTrack(tracks, selectedVideoId));
    }

    private String findErrorAlert(JsonBrowser jsonResponse) {
        JsonBrowser alerts = jsonResponse.get("alerts");

        if (!alerts.isNull()) {
            for (JsonBrowser alert : alerts.values()) {
                JsonBrowser alertInner = alert.get("alertRenderer");
                String type = alertInner.get("type").text();

                if ("ERROR".equals(type)) {
                    JsonBrowser textObject = alertInner.get("text");

                    String text;
                    if (!textObject.get("simpleText").isNull()) {
                        text = textObject.get("simpleText").text();
                    } else {
                        text = textObject.get("runs").values().stream()
                            .map(run -> run.get("text").text())
                            .collect(Collectors.joining());
                    }

                    return text;
                }
            }
        }

        return null;
    }

    private AudioTrack findSelectedTrack(List<AudioTrack> tracks, String selectedVideoId) {
        if (selectedVideoId != null) {
            for (AudioTrack track : tracks) {
                if (selectedVideoId.equals(track.getIdentifier())) {
                    return track;
                }
            }
        }

        return null;
    }

    private String extractPlaylistTracks(JsonBrowser playlistVideoList, List<AudioTrack> tracks,
                                         Function<AudioTrackInfo, AudioTrack> trackFactory) {
        JsonBrowser contents = playlistVideoList.get("contents");
        if (contents.isNull()) return null;

        final List<JsonBrowser> playlistTrackEntries = contents.values();
        for (JsonBrowser track : playlistTrackEntries) {
            JsonBrowser item = track.get("playlistVideoRenderer");

            JsonBrowser shortBylineText = item.get("shortBylineText");

            // If the isPlayable property does not exist, it means the video is removed or private
            // If the shortBylineText property does not exist, it means the Track is Region blocked
            if (!item.get("isPlayable").isNull() && !shortBylineText.isNull()) {
                String videoId = item.get("videoId").text();
                JsonBrowser titleField = item.get("title");
                String title = Optional
                    .ofNullable(titleField.get("simpleText").text())
                    .orElse(titleField.get("runs").index(0).get("text").text());

                String author = shortBylineText.get("runs").index(0).get("text").text();
                JsonBrowser lengthSeconds = item.get("lengthSeconds");
                long duration = Units.secondsToMillis(lengthSeconds.asLong(Units.DURATION_SEC_UNKNOWN));

                AudioTrackInfo info = new AudioTrackInfo(
                    title,
                    author,
                    duration,
                    videoId,
                    false,
                    "https://www.youtube.com/watch?v=" + videoId,
                    ThumbnailTools.extractYouTube(item, videoId));

                tracks.add(trackFactory.apply(info));
            }
        }

        JsonBrowser continuations = playlistVideoList.get("continuations")
            .index(0)
            .get("nextContinuationData");
        String continuationsToken;
        if (!continuations.isNull()) {
            continuationsToken = continuations.get("continuation").text();
            return continuationsToken;
        }

        return null;
    }
}

File: main/src/main/java/lavaplayer/source/youtube/DefaultYoutubePlaylistLoader.kt
package lavaplayer.source.youtube

import lavaplayer.tools.FriendlyException
import lavaplayer.tools.io.HttpClientTools
import lavaplayer.tools.io.HttpInterface
import lavaplayer.tools.json.JsonBrowser
import lavaplayer.track.*
import org.apache.http.client.methods.HttpPost
import org.apache.http.entity.StringEntity
import org.apache.http.util.EntityUtils
import java.io.IOException

class DefaultYoutubePlaylistLoader : YoutubePlaylistLoader {
    @Volatile
    override var playlistPageCount = 6

    override fun load(
        httpInterface: HttpInterface,
        playlistId: String,
        selectedVideoId: String?,
        trackFactory: AudioTrackFactory
    ): BasicAudioTrackCollection {
        val post = HttpPost(YoutubeConstants.BROWSE_URL)
        post.entity = StringEntity(YoutubeConstants.BROWSE_PLAYLIST_PAYLOAD.format(playlistId), "UTF-8")

        try {
            httpInterface.execute(post).use { response ->
                HttpClientTools.assertSuccessWithContent(response, "playlist response")
                HttpClientTools.assertJsonContentType(response)
                val playlist = JsonBrowser.parse(response.entity.content)
                return buildPlaylist(httpInterface, playlist, selectedVideoId, trackFactory)
            }
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    @Throws(IOException::class)
    private fun buildPlaylist(
        httpInterface: HttpInterface,
        playlist: JsonBrowser,
        selectedVideoId: String?,
        trackFactory: AudioTrackFactory
    ): BasicAudioTrackCollection {
        findErrorAlert(playlist)?.let { message ->
            throw FriendlyException(message, FriendlyException.Severity.COMMON, null)
        }

        val playlistName =
            playlist["header"]["playlistHeaderRenderer"]["title"]["runs"][0]["text"].text

        val playlistVideoList =
            playlist["contents"]["singleColumnBrowseResultsRenderer"]["tabs"][0]["tabRenderer"]["content"]["sectionListRenderer"]["contents"][0]["playlistVideoListRenderer"]

        val tracks: MutableList<AudioTrack> = ArrayList()
        var continuationsToken = extractPlaylistTracks(playlistVideoList, tracks, trackFactory)
        var loadCount = 0
        val pageCount = playlistPageCount

        // Also load the next pages, each result gives us a JSON with separate values for list html and next page loader html
        while (continuationsToken != null && ++loadCount < pageCount) {
            val post = HttpPost(YoutubeConstants.BROWSE_URL)
            post.entity = StringEntity(YoutubeConstants.BROWSE_CONTINUATION_PAYLOAD.format(continuationsToken), "UTF-8")

            httpInterface.execute(post).use { response ->
                HttpClientTools.assertSuccessWithContent(response, "playlist response")

                val continuation = JsonBrowser.parse(response.entity.content)
                val playlistVideoListPage = continuation["continuationContents"]["playlistVideoListContinuation"]

                continuationsToken = extractPlaylistTracks(playlistVideoListPage, tracks, trackFactory)
            }
        }

        return BasicAudioTrackCollection(
            playlistName!!,
            AudioTrackCollectionType.Playlist,
            tracks,
            findSelectedTrack(tracks, selectedVideoId)
        )
    }

    private fun findErrorAlert(jsonResponse: JsonBrowser): String? {
        val alerts = jsonResponse["alerts"]
        if (!alerts.isNull) {
            val textObject = alerts.values()
                .firstOrNull { it["alertRenderer"]["type"].text == "ERROR" }
                ?.let { it["alertRenderer"]["text"] } ?: return null

            return if (!textObject["simpleText"].isNull) {
                textObject["simpleText"].text
            } else {
                textObject["runs"].values()
                    .map { it["text"].text }
                    .joinToString("")
            }
        }

        return null
    }

    private fun findSelectedTrack(tracks: List<AudioTrack>, selectedVideoId: String?): AudioTrack? {
        if (selectedVideoId != null) {
            return tracks.firstOrNull { it.identifier == selectedVideoId }
        }

        return null
    }

    private fun extractPlaylistTracks(playlistVideoList: JsonBrowser, tracks: MutableList<AudioTrack>, trackFactory: AudioTrackFactory): String? {
        val contents = playlistVideoList["contents"].takeUnless { it.isNull }
            ?: return null

        for (track in contents.values()) {
            val video = track["playlistVideoRenderer"].cast<YouTubeVideoModel>()

            // If the isPlayable property does not exist, it means the video is removed or private
            // If the shortBylineText property does not exist, it means the Track is Region blocked
            if (video.isPlayable != null && video.author != null) {

                /* create the audio track. */
                val info = AudioTrackInfo(
                    title = video.title,
                    author = video.author!!,
                    length = video.length!!,
                    identifier = video.id,
                    uri = "https://www.youtube.com/watch?v=${video.id}",
                    artworkUrl = video.thumbnail,
                )

                tracks.add(trackFactory.create(info))
            }
        }

        val continuations = playlistVideoList["continuations"][0]["nextContinuationData"]
        if (!continuations.isNull) {
            return continuations["continuation"].text
        }

        return null
    }
}
